### DISCUSSION TITLE
Deploying ganache-local w/brownie: VM Exception while processing transaction: invalid opcode

### DISCUSSION THREAD

**Discussion by GratefulDave** [smartcontractkit/full-blockchain-solidity-course-py#422] [Q&A] [answered]
Thanks for an amazing class Patrick. I taught myself to code 3 years ago and wish I did it 33 years ago! :)

# Issue 
I am going through brownie simple storage part of this course. The initial deploy,.py works properly, but I keeping getting some variation of invalid opcode and the transaction is reverting when I try to deploy to ganache-local. I am able to run the tests in this project to development/ganache-cli, but fail with ganache-local and Rinkeby. 

I am experiencing the same issues as I am with my own code base. I decided to test with a clone of **this** repo as 'bss' and replacing .env files with my info and am experiencing the same issues. I'm on MacOS.

# Steps to reproduce & Troubleshooting

1. brownie compile -a.  Compiles successfully. 
```Compiling contracts...
  Solc version: 0.6.12
  Optimizer: Enabled  Runs: 200
  EVM Version: Istanbul
Generating build data...
 - smartcontractkit/chainlink-brownie-contracts@1.1.1/AggregatorInterface
 - smartcontractkit/chainlink-brownie-contracts@1.1.1/AggregatorV2V3Interface
 - smartcontractkit/chainlink-brownie-contracts@1.1.1/AggregatorV3Interface
 - smartcontractkit/chainlink-brownie-contracts@1.1.1/SafeMathChainlink
 - FundMe
 - MockV3Aggregator
 Project has been compiled. Build artifacts saved at /Users/davidandrews/Documents/Projects/Web3/bss/build/contracts
```
2. Deploy ganache ui.
3. Execute `brownie run scripts/deploy.py --network ganache-local `
```
 ~/Documents/Projects/Web3/bss (main●●)$ brownie run scripts/deploy.py --network ganache-local 
Brownie v1.17.1 - Python development framework for Ethereum

BssProject is the active project.

Running 'scripts/deploy.py::main'...
The active network is ganache-local
Deploying Mocks...
Transaction sent: 0xe7b501480f70836256dbdd37a159100198d1422897117777e33588d1b2e1a566
  Gas price: 20.0 gwei   Gas limit: 540226   Nonce: 0
  MockV3Aggregator.constructor confirmed   Block: 1   Gas used: 491115 (90.91%)
  MockV3Aggregator deployed at: 0xF5F2B13A72901f1621d2CD9b382c1c513ac0724D

Mocks Deployed!
Transaction sent: 0xc5a26f42997e9c648ee4d12df1805746cd3ca6508c315747f9b836ce5fb39516
  Gas price: 20.0 gwei   Gas limit: 528696   Nonce: 1
  FundMe.constructor confirmed   Block: 2   Gas used: 480633 (90.91%)
  FundMe deployed at: 0xEDa314a434Ef759d85F60a54804FdfdbEB26F179

Contract deployed to 0xEDa314a434Ef759d85F60a54804FdfdbEB26F179
```
4. Execute: ` brownie run scripts/fund_and_withdraw.py --network ganache-local`
```Brownie v1.17.1 - Python development framework for Ethereum

BssProject is the active project.

Running 'scripts/fund_and_withdraw.py::main'...
25000000000000000
The current entry fee is 25000000000000000
Funding
Transaction sent: 0xffd686e74b439f0afff97ba5ab77ce31a48a66a6f2710d123b3cf37bc8b8bade
  Gas price: 20.0 gwei   Gas limit: 94207   Nonce: 2
  FundMe.fund confirmed   Block: 3   Gas used: 85643 (90.91%)

  File "brownie/_cli/run.py", line 50, in main
    return_value, frame = run(
  File "brownie/project/scripts.py", line 103, in run
    return_value = f_locals[method_name](*args, **kwargs)
  File "./scripts/fund_and_withdraw.py", line 23, in main
    withdraw()
  File "./scripts/fund_and_withdraw.py", line 18, in withdraw
    fund_me.withdraw({"from": account})
  File "brownie/network/contract.py", line 1625, in __call__
    return self.transact(*args)
  File "brownie/network/contract.py", line 1498, in transact
    return tx["from"].transfer(
  File "brownie/network/account.py", line 644, in transfer
    receipt, exc = self._make_transaction(
  File "brownie/network/account.py", line 727, in _make_transaction
    raise VirtualMachineError(e) from None
  File "brownie/exceptions.py", line 121, in __init__
    raise ValueError(str(exc)) from None
ValueError: Gas estimation failed: 'execution reverted: VM Exception while processing transaction: invalid opcode'. This transaction will likely revert. If you wish to broadcast, you must set the gas limit manually.
```
5. Add gas_limit param to `fund_me.withdraw({"from": account})` -> `fund_me.withdraw({"from": account, 'gas_limit': 6721975})`
```
Brownie v1.17.1 - Python development framework for Ethereum

BssProject is the active project.

Running 'scripts/fund_and_withdraw.py::main'...
25000000000000000
The current entry fee is 25000000000000000
Funding
Transaction sent: 0x996f97219b9c8cf1dd3f88753adf4c94335d27d1784972b976c550cea5f13c6d
  Gas price: 20.0 gwei   Gas limit: 61207   Nonce: 3
  FundMe.fund confirmed   Block: 4   Gas used: 55643 (90.91%)

  File "brownie/_cli/run.py", line 50, in main
    return_value, frame = run(
  File "brownie/project/scripts.py", line 103, in run
    return_value = f_locals[method_name](*args, **kwargs)
  File "./scripts/fund_and_withdraw.py", line 24, in main
    withdraw()
  File "./scripts/fund_and_withdraw.py", line 19, in withdraw
    fund_me.withdraw({"from": account, 'gas_limit': 6721975})
  File "brownie/network/contract.py", line 1625, in __call__
    return self.transact(*args)
  File "brownie/network/contract.py", line 1498, in transact
    return tx["from"].transfer(
  File "brownie/network/account.py", line 644, in transfer
    receipt, exc = self._make_transaction(
  File "brownie/network/account.py", line 752, in _make_transaction
    exc = VirtualMachineError(e)
  File "brownie/exceptions.py", line 121, in __init__
    raise ValueError(str(exc)) from None
ValueError: Execution reverted during call: 'execution reverted: VM Exception while processing transaction: invalid opcode'. This transaction will likely revert. If you wish to broadcast, include `allow_revert:True` as a transaction parameter.
```
6. I've also tried deleting build artifacts & redeploying ganache ui.
 
# Environment
Brownie `v.1.17.1`
Ganache CLI `v6.12.2 (ganache-core: 2.13.2)`
Node `v16.8.0`
Npm `v7.21.0`
Truffle `v5.0.22`

The only part of the code that has been changed from the repo is adding the gas_limit. Still happy to share the code if request. For the life of me, I can't figure what invalid opcode is causing this to revert . . . Any help would be greatly appreciated!

Thanks in advance!
@gratefuldave

**Comment by cromewar** [comment]
Hello @GratefulDave first of all thanks for the detailed description of your error, it seems the application is reverting because  you are not using Mocks to get the price feed, so if you are requesting the price feed directly to chainlink.
ganache-local should use local accounts, so please check your helpful_scripts and see you have added it to local chains:

![image](https://user-images.githubusercontent.com/20199180/142804955-81d109d1-dd02-479f-ae0a-a29be8e9300f.png)
![image](https://user-images.githubusercontent.com/20199180/142804975-7bdc7eb9-1cbb-44d5-9836-a1142f7c5259.png)

your deploy script should deploy mocks also for ganache-local as you can see:


```Python
def deploy_fund_me():
    account = get_account()
    if network.show_active() not in LOCAL_BLOCKCHAIN_ENVIRONMENTS:
        price_feed_address = config["networks"][network.show_active()][
            "eth_usd_price_feed"
        ]
    else:
        deploy_mocks()
        price_feed_address = MockV3Aggregator[-1].address

    fund_me = FundMe.deploy(
        price_feed_address,
        {"from": account},
        publish_source=config["networks"][network.show_active()].get("verify"),
    )
    print(f"Contract deployed to {fund_me.address}")
    return fund_me
 
```
Regarding to Rinkeby you must have some eth to pay the transaction, so you must use a faucet to get some eth.

Also remember `fundme_and_withdraw` takes the last deploy, so you need to deploy the contract in both ganache-local and Rinkeby first, before running this:  `brownie run scripts/fund_and_withdraw.py --network ganache-local`.

Please let us know if this info was helpful to you.

Best regards, Cromewar.

**Reply by GratefulDave** [reply]
Hi Cromewar,

Thanks for your reply. 
1. I have enough ETH for Rinkeby from the faucets.
2. If you look at item '#3' in my description above. I am deploying mocks. I had my own code, but cloned repo to see if I could get the repo code to work with my environment. 

Here is my environment:

Brownie `v.1.17.1`
Ganache CLI `v6.12.2 (ganache-core: 2.13.2)`
Node `v16.8.0`
Npm `v7.21.0`
Truffle `v5.0.22`

Just tried again . . . still getting same errors.

**Reply by cromewar** [reply]
Hello there, I read the link you provided and it seems you would like to use Kovan instead of Rinkeby to test the code, however ganache-local should have worked anyway with mocks. @PatrickAlphaC could you give us a hand on this one?

**Reply by GratefulDave** [reply]
The linkI provided was automatically inserted by the Github editor for some reason. I was referring to point number 3 in my description.

**Reply by GratefulDave** [reply]
And I am using deploy_mocks in my code.

**Reply by cromewar** [reply]
Hello @GratefulDave I was trying to do research on your problem but I was unable to reproduce the problem meanwhile @PatrickAlphaC give a check to this problem I'll suggest you to try one of this things:

1. Use a python virtualenv for the project.
2. Maybe something is messing up with your brownie installation, may be reinstalling it would help (also deleting .brownie folder on your home path before installing again).
3. Just remove ganache-local from brownie and add it again.

**Comment by GratefulDave** [comment]
Hi Vasiliy,

I am using a virtualenv. I’ve tried removing and reinstalling brownie and pipx, but have not tried deleting the .brownie or removing the network from brownie and adding it again.

What’s odd is that everything else works. I’ll try these solutions and let you know how it goes. 

Thank you!

@gratefuldave

**Comment by GratefulDave** [comment]
1. I am using a virtualenv.
2. I removed pipx and brownie.
3. I removed the brownie fold from ~/.local/bin directory.
4. I removed and added the ganache-local network which is for me https://127.0.0.1:7545 and chainid=5777

Still experiencing the same behavior. :/

**Reply by cromewar** [reply]
Well that wasn't helpful, sorry for that. I don't know why the price feed is reverting your transaction, I'll do further research to look if I can find something else that might help.

**Comment by PatrickAlphaC** [comment]
Hmm... any chance you can upload what you have to a github repo so we can reproduce what you have? 

Thanks for making this issue, and thanks for the flag @cromewar !

**Comment by GratefulDave** [comment]
I uploaded it to a private repo and sent you an invite. I had a private key in the .env with a small balance - which I cleared. There’s also one worthless NFT. Lol.

**Reply by cromewar** [reply]
I hope Patrick would be able to check it, just please remove the private key even if the account has not founds is a bad practice sharing them, best regards, Cromewar.

**Comment by AnneCh** [comment]
Hi everyone,

Thank you @GratefulDave for tagging your post on mine, I'll follow how things are evolving. I've spent today trying everything that was suggested in this thread and just like you, keep having the same error showing.

**Comment by GratefulDave** [comment]
@PatrickAlphaC here's the repo: https://github.com/GratefulDave/brownie-fund-me.git.

**Comment by cromewar** [comment]
Hello @GratefulDave I just cloned your repo and tried to reproduce the error, but unfortunately I couldn't as everything worked just fine using ganache-local, so I think something is messing up with your configuration, as you can see bellow it worked just fine:
![image](https://user-images.githubusercontent.com/20199180/143918522-8092f02b-dd8e-4f4c-a347-d4ef4bc2564c.png)

So I have some recommendations for you:

1. first of all, remove the ganache-local from brownie: `brownie networks remove ganache-local`.
2. open a new instance of ganache-ui and configure the network ID to 1337

![image](https://user-images.githubusercontent.com/20199180/143918724-fd03b0f0-7197-4e4c-9958-8067ff887460.png)

3. pay attention to the network config, specially to the RPC Server (yours is going to be different as I'm using Linux server)

![image](https://user-images.githubusercontent.com/20199180/143918852-d32ee4dd-c85e-4f5d-8601-cc3155a09dab.png)

4. add the again ganache-local to brownie using: `brownie networks add Ethereum ganache-local host=http://172.17.224.1:7545 chainid=1337` (the host would be whatever you have on RPC server above, remember "http" must be lowercase).
5. Without closing ganache, try to deploy again.

I hope this might help.

**Reply by GratefulDave** [reply]
Thanks for the reply. I tried all of that - same error. deploy.py works, fund_and_withdraw.py does not. Now I am getting this:

`ValueError: Gas estimation failed: 'execution reverted: VM Exception while processing transaction: invalid opcode'. This transaction will likely revert. If you wish to broadcast, you must set the gas limit manually.

When I add a gas_limit like this to the withdraw() function `fund_me.withdraw({"from": account, 'gas_limit': 6721975}` it throws:

`ValueError: Execution reverted during call: 'execution reverted: VM Exception while processing transaction: invalid opcode'. This transaction will likely revert. If you wish to broadcast, include allow_revert:True as a transaction parameter.
`

When I add `allow_revert:True` as a transaction parameter like this `fund_me.withdraw({"from": account, 'gas_limit': 6721975, 'allow_revert': True})` it throws the following error:

`VirtualMachineError: invalid opcode: invalid opcode`

@AnneCh since you and I are having the same exact error - you may want to try this solution and lmk if you have any success.

@cromewar : How do I determine which opcode is invalid?

**Reply by cromewar** [reply]
Just run `fund_and_withdraw` and still don't getting the error: 

![image](https://user-images.githubusercontent.com/20199180/143928288-cf2b9f9d-b742-414d-bcac-e6d4934e5479.png)
![image](https://user-images.githubusercontent.com/20199180/143928346-dd6d1e42-14fb-4a33-aff0-3facdcab50ac.png)

Also I made a little change on the script as follows:

Basically imported deploy script just to be able to access to FundMe[-1] at the execution but I don't think is relevant.
```python

from brownie import FundMe
from scripts.helpful_scripts import get_account, deploy_mocks
from scripts.deploy import deploy_fund_me


def fund():
    fund_me = FundMe[-1]
    account = get_account()
    entrance_fee = fund_me.getEntranceFee()
    print(entrance_fee)
    print(f"The current entry fee is {entrance_fee}")
    print("Funding")
    fund_me.fund({"from": account, "value": entrance_fee})


def withdraw():
    fund_me = FundMe[-1]
    account = get_account()
    # fund_me.withdraw({"from": account, 'gas_limit': 6721975})
    fund_me.withdraw({"from": account})


def main():
    deploy_fund_me()
    fund()
    withdraw()

```

Is strange you get a revert on the transaction as the accounts have enough gas, I'll keep trying to reproduce the error.

**Reply by GratefulDave** [reply]
I ran the code through brownie console. I set `tx = fund_me.withdraw({"from": account, 'gas_limit': 6721975})`.  When I enter `tx.error()` this is the response:

```
Trace step 55, program counter 653:
  File "contracts/FundMe.sol", line 68, in FundMe.withdraw:    
    }

    function withdraw() public payable onlyOwner {
        msg.sender.transfer(address(this).balance);

        for (
            uint256 funderIndex = 0;
```
I'm using the following constructor:
```
constructor(address _priceFeed) public {
        priceFeed = AggregatorV3Interface(_priceFeed);
        owner = msg.sender;
    }
```
And onlyOwner modifier . . . 
```
modifier onlyOwner() {
        require(msg.sender == owner);
        _;
    }
```
It appears to be failing on address(this).balance. 
Here's the `tx.traceback()` response:

```
Traceback for '0xcf643d1a6bfc0a8ca0d45761147a8ac657afea545be392cb3b3d3c2b394c4245':
Trace step 55, program counter 653:
  File "contracts/FundMe.sol", line 68, in FundMe.withdraw:    
    msg.sender.transfer(address(this).balance);
```

**Reply by GratefulDave** [reply]
Here's `tx.call_trace()`:

```
Call trace for '0xc0581cde1d13050278bde51579f019ea898b951b3b64b7d62a5951f3a4e5d984':
Initial call cost  [6721568 gas]
FundMe.withdraw  0:56  [407 gas]
```
I've attached the output from tx.trace as well. 
[trace.txt](https://github.com/smartcontractkit/full-blockchain-solidity-course-py/files/7620965/trace.txt)

**Reply by cromewar** [reply]
I did more research on this and I couldn't reproduce the error as everything worked just fine using the code you provided @GratefulDave I would like to suggest to post this issue on official brownie github, may be the developers will have a better understanding about why this is happening.

**Reply by GratefulDave** [reply]
I don't have the expertise - yet - to understand what opcode is failing. If you or @PatrickAlphaC could explain what the opcode is and why it is failing that might help @AnneCh and I diagnose the issue.

**Comment by AnneCh** [comment]
Hi!

So, being stuck just like @GratefulDave , I've followed the advice from @cromewar by changing my Ganache network ID. Now my error is different, brownie is somehow not taking the entrance fee. I've tried with Dave's codes, as well as with the original files from Patrick on his repo, and the error comes back the exact same every time:

```
PS C:\Users\chret\Documents\demo\brownie_fund_me> brownie run scripts/fund_and_withdraw.py --network ganache-local
INFO: Could not find files for the given pattern(s).
Brownie v1.17.1 - Python development framework for Ethereum

BrownieFundMeProject is the active project.

Running 'scripts\fund_and_withdraw.py::main'...
0
The current entry fee is 0
Funding
  File "C:\Users\chret\AppData\Local\Programs\Python\Python39\lib\site-packages\brownie\_cli\run.py", line 50, in main
    return_value, frame = run(
  File "C:\Users\chret\AppData\Local\Programs\Python\Python39\lib\site-packages\brownie\project\scripts.py", line 103, in run
    return_value = f_locals[method_name](*args, **kwargs)
  File ".\scripts\fund_and_withdraw.py", line 22, in main
    fund()
  File ".\scripts\fund_and_withdraw.py", line 12, in fund
    fund_me.fund({"from": account, "value": entrance_fee})
  File "C:\Users\chret\AppData\Local\Programs\Python\Python39\lib\site-packages\brownie\network\contract.py", line 1625, in __call__
    return self.transact(*args)
  File "C:\Users\chret\AppData\Local\Programs\Python\Python39\lib\site-packages\brownie\network\contract.py", line 1498, in transact
    return tx["from"].transfer(
  File "C:\Users\chret\AppData\Local\Programs\Python\Python39\lib\site-packages\brownie\network\account.py", line 652, in transfer
    receipt, exc = self._make_transaction(
  File "C:\Users\chret\AppData\Local\Programs\Python\Python39\lib\site-packages\brownie\network\account.py", line 735, in _make_transaction
    raise VirtualMachineError(e) from None
  File "C:\Users\chret\AppData\Local\Programs\Python\Python39\lib\site-packages\brownie\exceptions.py", line 125, in __init__
    raise ValueError(str(exc)) from None
ValueError: Gas estimation failed: 'execution reverted: VM Exception while processing transaction: revert You need to spend more ETH!'. This transaction will likely revert. If you wish to broadcast, you must set the gas limit manually.
```

I've modified once again the ganache network to try to get back to the original error, deleted it from Brownie, re-added it, three times, but the abovementioned error remains the same.

**Reply by AnneCh** [reply]
note that running deploy.py works perfectly

**Reply by GratefulDave** [reply]
I had the same error prior to that. Try using this for withdraw: `fund_me.withdraw({"from": account, 'gas_limit': 6721975, 'allow_revert': True})` and tell me if you get the same error.

**Reply by AnneCh** [reply]
I saw your post where you had added this code, and tried it as well. Unfortunately, same error message, it's still not catching the entry fee for some reason. I've opened an issue on the eth-brownie/brownie page, hopefully someone will have some answer for us : [1358](https://github.com/eth-brownie/brownie/issues/1358)

**Comment by cromewar** [comment]
@GratefulDave I was able to reproduce de error adding this `fund_me.withdraw({"from": account, 'gas_limit': 6721975, 'allow_revert': True})` but I also manage to solve it, follow this steps.

On the terminal type:
1. `rm -rf build` (this will remove completely all the previous builds).
2. Change the fund_and_withdraw as follows:
```Python
def fund():
    fund_me = FundMe[-1]
    account = get_account()
    entrance_fee = fund_me.getEntranceFee()
    print(entrance_fee)
    print(f"The current entry fee is {entrance_fee}")
    print("Funding")
    tx = fund_me.fund({"from": account, "value": entrance_fee})
    tx.wait(1)


def withdraw():
    fund_me = FundMe[-1]
    account = get_account()
    # fund_me.withdraw({"from": account, 'gas_limit': 6721975})
    tx = fund_me.withdraw({"from": account, "gas_limit": 6721975, "allow_revert": True})
    tx.wait(1)

```

and try again.

Best regards, Cromewar.

**Reply by AnneCh** [reply]
@GratefulDave let me know if that worked for you, it didn't for me

**Reply by GratefulDave** [reply]
@cromewar I still get the same error. I wish I understood what the specific opcode issue was.

**Reply by cromewar** [reply]
Well I'm lost then, even I was able to reproduce the error for me it just was enough to delete the build folder as I post above. I hope @AnneCh could get some info on the brownie repo.

**Reply by GratefulDave** [reply]
@cromewar Can you share your requirements.txt file?

**Comment by cromewar** [comment]
Here you have.

```Python
aiohttp==3.7.4.post0
asttokens==2.0.4
async-timeout==3.0.1
attrs==21.2.0
base58==2.1.0
bitarray==1.2.2
black==21.9b0
certifi==2021.5.30
chardet==4.0.0
charset-normalizer==2.0.6
click==8.0.1
cytoolz==0.11.0
dataclassy==0.10.4
eip712==0.1.0
eth-abi==2.1.1
eth-account==0.5.5
eth-brownie==1.16.4
eth-event==1.2.3
eth-hash==0.3.2
eth-keyfile==0.5.1
eth-keys==0.3.3
eth-rlp==0.2.1
eth-typing==2.2.2
eth-utils==1.10.0
execnet==1.9.0
hexbytes==0.2.2
hypothesis==6.21.6
idna==3.2
inflection==0.5.0
iniconfig==1.1.1
ipfshttpclient==0.8.0a2
jsonschema==3.2.0
lazy-object-proxy==1.6.0
lru-dict==1.1.7
multiaddr==0.0.9
multidict==5.1.0
mypy-extensions==0.4.3
mythx-models==1.9.1
netaddr==0.8.0
packaging==21.0
parsimonious==0.8.1
pathspec==0.9.0
platformdirs==2.3.0
pluggy==1.0.0
prompt-toolkit==3.0.20
protobuf==3.18.0
psutil==5.8.0
py==1.10.0
py-solc-ast==1.2.9
py-solc-x==1.1.0
pycryptodome==3.10.1
Pygments==2.10.0
pygments-lexer-solidity==0.7.0
PyJWT==1.7.1
pyparsing==2.4.7
pyrsistent==0.18.0
pytest==6.2.5
pytest-forked==1.3.0
pytest-xdist==1.34.0
python-dateutil==2.8.1
python-dotenv==0.16.0
pythx==1.6.1
PyYAML==5.4.1
regex==2021.8.28
requests==2.26.0
rlp==1.2.0
semantic-version==2.8.5
six==1.16.0
sortedcontainers==2.4.0
toml==0.10.2
tomli==1.2.1
toolz==0.11.1
tqdm==4.62.3
typing-extensions==3.10.0.2
urllib3==1.26.6
varint==1.0.2
vvm==0.1.0
vyper==0.2.16
wcwidth==0.2.5
web3==5.23.1
websockets==9.1
wrapt==1.12.1
yarl==1.6.3

```

**Reply by GratefulDave** [reply]
Still no success. I set the gas limit to the max gas limit in ganache UI. I noticed that the transaction that is reverting has used the entire gas limit. Is that the likely cause for the contract to revert? If so, any ideas on how to solve?

**Comment by AnneCh** [comment]
Update, I've run the following commands:
```
$ npm uninstall ganache-cli --global
$ npm install ganache@alpha --global
```

and now the entry fee is grabbing the correct number, the contract is called and shows on my Ganache transaction page, but the withdraw function still doesn't work. So, I fixed the problem I had created when changing the chainID on Ganache, back to the original problem where brownie tells me 
```
 File "C:\Users\chret\AppData\Local\Programs\Python\Python39\lib\site-packages\brownie\_cli\run.py", line 50, in main
    return_value, frame = run(
  File "C:\Users\chret\AppData\Local\Programs\Python\Python39\lib\site-packages\brownie\project\scripts.py", line 103, in run
    return_value = f_locals[method_name](*args, **kwargs)
  File ".\scripts\fund_and_withdraw.py", line 23, in main
    withdraw()
  File ".\scripts\fund_and_withdraw.py", line 18, in withdraw
    fund_me.withdraw({"from": account})
  File "C:\Users\chret\AppData\Local\Programs\Python\Python39\lib\site-packages\brownie\network\contract.py", line 1625, in __call__
    return self.transact(*args)
  File "C:\Users\chret\AppData\Local\Programs\Python\Python39\lib\site-packages\brownie\network\contract.py", line 1498, in transact
    return tx["from"].transfer(
  File "C:\Users\chret\AppData\Local\Programs\Python\Python39\lib\site-packages\brownie\network\account.py", line 652, in transfer
    receipt, exc = self._make_transaction(
  File "C:\Users\chret\AppData\Local\Programs\Python\Python39\lib\site-packages\brownie\network\account.py", line 735, in _make_transaction
    raise VirtualMachineError(e) from None
  File "C:\Users\chret\AppData\Local\Programs\Python\Python39\lib\site-packages\brownie\exceptions.py", line 125, in __init__
    raise ValueError(str(exc)) from None
ValueError: Gas estimation failed: 'execution reverted: VM Exception while processing transaction: invalid opcode'. This transaction will likely revert. If you wish to broadcast, you must set the gas limit manually.
```

@GratefulDave I'll update my post on eth-brownie/brownie. Fingers crossed someone will help us out, I really want to keep learning through this tutorial! (making zombies in the meantime!)

**Comment by AnneCh** [comment]
@GratefulDave I found a solution on my part, change the gas limit on Ganache itself, for which you have to create a new workspace. I've described my process on [1358](https://github.com/eth-brownie/brownie/issues/1358). 
Let me know if that works for you!! 5 days of being stuck are OVER.

**Reply by cromewar** [reply]
Your are awesome I hope this also works for Dave. So you changes the gas limit on a new ganache UI instance right?

If there are more users with this problem I'll suggest add your solution to the course index troubleshoot.

**Reply by AnneCh** [reply]
yes exactly, I didn't expect it to work but it did! 
Dave said he tried and it didn't work for him...

**Reply by cromewar** [reply]
@GratefulDave if your have any updates on this, let us know.

**Comment by GratefulDave** [comment]
This is the only answer I am getting from Ethereum.stackexcahnge so far . . . 

getPrice() isn't returning a number you want from the mock, somewhere in the vicinity of 2B. Think this is a bug with the Ganache implementation- performing the calculation (minimumUSD * precision) / price in getEntranceFee() gives you a number less than 1- and, since Solidity can't handle floats, Solidity simply sees it as a 0, and the whole thing errors out.

**Reply by cromewar** [reply]
humm if this is True, try to change the  return statement of get entrance fee and use a burned value you choose, just for testing purposes.

**Reply by GratefulDave** [reply]
I'm sorry, I'm not understanding what you mean by a 'burned value'.  If you are referring to changing the return value of the following function:
```
function getEntranceFee() public view returns (uint256) {
        // mimimumUSD
        uint256 mimimumUSD = 50 * 10 ** 18;
        uint256 price = getPrice();
        uint256 precision = 1 * 10 ** 18;
        return (mimimumUSD * precision) / price;
    }
```
I've tried:
- `1` it throws `Gas estimation failed: 'execution reverted: VM Exception while processing transaction: revert You need to spend more ETH!'. This transaction will likely revert. If you wish to broadcast, you must set the gas limit manually.`
- `minimumUSD` it throws `ValueError: sender doesn't have enough funds to send tx. The upfront cost is: 50001884140000000000 and the sender's account only has: 48393181040000000000`
- `price` it throws `ValueError: sender doesn't have enough funds to send tx. The upfront cost is: 2000001884140000000000 and the sender's account only has: 48383694580000000000`
- `precision` it throws `VirtualMachineError: invalid opcode: invalid opcode`

**Reply by GratefulDave** [reply]
I was able to get this script to work properly on an old, but clean MacBook Pro. I would like to be able to run this on my existing computer, but have no idea what could be causing the conflict. Can you point me in the right direction (or resources)?

**Reply by cromewar** [reply]
Does your Mac has a M1 chip? as far as I know they have struggles with web3 libraries. If not, I pretty much would say some libraries you have on your mac are conflicting with the ones needed for web3 development. I would pretty much suggest to purge completely python from your MAC and install it again with fresh libraries.

**Comment by JasonDHubbard** [comment, marked as answer]
Hi. 

I was able to solve the problems mentioned above in my environment.

### Resolution
Both the "ValueError: Gas estimation failed: 'execution reverted: VM Exception while processing transaction: invalid opcode'" error and the getEntranceFee() return value problem was _**resolved in my local environment by changing the "Hard Fork" property in the Ganache UI app from "Petersburg" to "Muir Glacier".**_ In the Ganachi UI app, go to Settings > Chain > HARDFORK. 

I do not know why the "Petersburg" hardfork does not work. 

fund_and_withdraw.py also required "gas_limit" being passed to fund_me.withdraw. If I remember right, that was due to changes in new versions of web.py. It is possible you may not need to pass in the gas_limit. The change...

`fund_me.withdraw({"from": account, "gas_limit": 6721975})`

### My/our tests...

Before I started troubleshooting I did a diff on Patrick's code vs mine. They were the same in all ways that mattered with the exception of the "fund_me.withdraw()" statement above.  

**A friend and I were able to replicate the problem in both of our local environments:**
a) I ran my fund_and_withdraw.py with "Petersburg" as the hardfork. I received opcode error message and the wrong value from getEntranceFee (i.e. 2.5 * 10 power 24). 
b) I ran the script with the "Muir Glacier" hardfork which did not return the opcode error message and did return the right value for getEntranceFee (i.e. 2.5 * 10 power 16)
c) I ran my friend's fund_and_withdraw.py under "Petersburg" and "Muir Glacier" and had the same results as I did in (a) and (b).
d) My friend ran his fund_and_withdraw.py under "Petersburg" and "Muir Glacier" and had the same results as I did in (a) and (b). 

Note that my friend and I both cleaned out our "build > contracts" contract files and "build > deployment" deployment files every time we switched hardforks. This meant redeploying the contracts before running fund_and_withdraw.py again.

**For what it's worth, my environment and my friend's differed only slightly:**
a) Both Ganache UI at 2.5.4
b) My Brownie 1.17.2; his 1.17.1 (command 'brownie --version') 
b) My Python 3.9.9; his 3.9.0 (command 'python --version')
c) My web3.py 5.25.0; his 5.24.0 (command 'web3.api' in brownie console)

Hope this helps :)
Jason

**Reply by cromewar** [reply]
You are just amazing @JasonDHubbard

**Reply by JasonDHubbard** [reply]
@cromewar No, you :D

**Reply by PatrickAlphaC** [reply]
Awesome work!!

**Reply by RohanNero** [reply]
Thank you @JasonDHubbard! Never using the Petersburg hardfork again lol

**Comment by strudelPie** [comment]
I know this has been closed but **none** of the above solutions have worked for me :/ Really want to continue on with this amazing tutorial but cannot find a solution

**Reply by PatrickAlphaC** [reply]
Can you follow up with a new question with as much info as possible?

**Reply by cromewar** [reply]
Yes please @strudelPie, post here screenshots or any relevant info.

**Reply by strudelPie** [reply]
@PatrickAlphaC @cromewar Thanks very much in advance to you both for taking a look. Great course and support with issues on top. Much appreciated. New question raised at #901
